project/DistributionPackage.scala

```diff
@@ -471,20 +482,66 @@ object DistributionPackage {
   /** Returns the argument specifying the path of the project to run.
     *
     * It will be the argument following `--in-project`, `--run` or `--compile`.
+    *
+    * @param root the root of the engine distribution
+    * @return index of the replace argument (or -1)
     */
-  private def findProjectPath(args: Seq[String]): Option[String] = {
-    def findArg(name: String): Option[String] = {
+  private def findProjectPath(
+    root: File,
+    args: Seq[String]
+  ): (Int, java.io.File, String) = {
+    def findArg(name: String): Option[(Int, String)] = {
       val location = args.indexOf(name)
       if (location >= 0 && location + 1 < args.size) {
-        Some(args(location + 1))
+        Some((location + 1, args(location + 1)))
       } else {
         None
       }
     }
 
-    findArg("--in-project")
+    val indexPath = findArg("--in-project")
       .orElse(findArg("--run"))
       .orElse(findArg("--compile"))
+    if (indexPath.isEmpty) {
+      return (-1, null, null)
+    }
+
+    val index = indexPath.orNull._1
+    val path  = indexPath.orNull._2
+
+    val runnerJar = root / "component" / "engine-runner.jar"
+    if (!runnerJar.exists()) {
+      throw new IllegalStateException("Cannot find " + runnerJar)
+    }
+    val slf4jJar = root / "component" / "slf4j-api-2.0.16.jar"
+    if (!slf4jJar.exists()) {
+      throw new IllegalStateException("Cannot find " + slf4jJar)
+    }
+    val l = new java.net.URLClassLoader(
+      Array(
+        runnerJar.toURI().toURL(),
+        slf4jJar.toURI().toURL()
+      ),
+      Class.forName("scala.Tuple2").getClassLoader()
+    )
+    try {
+      val utils = l.loadClass("org.enso.runner.Utils")
```

Comment:
Revolutionary change. Let `sbt` code (usually hacky, untestable, written in Scala) call into properly tested code of `enso-runner`. Guarantees consistency.

The `engine-runner.jar` & co. has to exist (but they do as `sbt buildEngineDistribution` has already finished). The class is loaded by a temporary `ClassLoader`. E.g. it always loads new class. The class and its loader are no longer referenced by the JVM once this method is over, e.g. they can be GCed.

Reply:
Seems like a good idea, as long as it works.